torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -1196,6 +1213,329 @@ def call_function(
         return super().call_function(tx, args, kwargs)
 
 
+class CannotConvertRangeToHigherOrder(TorchDynamoException):
+    pass
+
+
+class RangeHigherOrderVariable(TorchHigherOrderOperatorVariable):
+    """
+    The general idea is that for a simple for loop like
+
+    ```
+    for i in range(10):
+        x = x + 1
+    ```
+
+    We can convert the loop body to a function and the loop itself
+    to a series of calls:
+
+    ```
+    def loop_body(local_0, local_1, local_...):
+        local_1 = local_1 + 1
+        return (local_0, local_1, local_...)
+
+    (i, x) = loop_body(0, x)
+    (i, x) = loop_body(1, x)
+    (i, x) = loop_body(2, x)
+    # ... 7 more times
+    ```
```

Comment:
Do we handle the situation where a new local gets constructed in the body?

Reply:
In CPython, the new local gets "hoisted" out. So yes. What happens is before the loop it will be unset, but already "declared" (using these terms loosely as Python has no declarations).